### PR TITLE
Charlesmchen/account attributes vs pin

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Util/OWS2FAManager.m [hunk 1]
```diff
@@ -65,16 +65,30 @@ - (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
...
+{
+    [self.dbConnection removeObjectForKey:kOWS2FAManager_PINKey inCollection:kOWS2FAManager_Collection];
+
+    [[NSNotificationCenter defaultCenter] postNotificationNameAsync:NSNotificationName_2FAStateDidChange
+                                                             object:nil
+                                                           userInfo:nil];
 }
 
-- (void)setIs2FAEnabled:(BOOL)value
+- (void)set2FAEnabledWithPin:(NSString *)pin
```
Comment (michaelkirk) [inline SignalServiceKit/src/Util/OWS2FAManager.m]: I had to make similar changes in #3085 - can you use those?

**Comment by charlesmchen** [comment]
@michaelkirk I'm going to wait until #3085 is merged before moving ahead with this.

File: SignalServiceKit/src/Util/OWS2FAManager.m (hunk 1 above)
Comment (charlesmchen) [reply SignalServiceKit/src/Util/OWS2FAManager.m]: Rebased and reconciled.  I drew from both but crucially we now have a single piece of state - a "registration lock" pin.  If it's set, "registration lock" is enabled.  Less state = less surface area for bugs.

**Comment by charlesmchen** [comment]
227
3,187 761 signalapp/Signal-iOS
 Code  Issues 331  Pull requests 30  Projects 0  Wiki  Insights  Settings

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

File: SignalServiceKit/src/Util/OWS2FAManager.m [hunk 2]
```diff
@@ -71,16 +70,19 @@ - (instancetype)initWithStorageManager:(TSStorageManager *)storageManager
...
+{
+    return [self.dbConnection objectForKey:kOWS2FAManager_PinCode inCollection:kOWS2FAManager_Collection];
+}
+
 - (BOOL)is2FAEnabled
 {
-    return [self.dbConnection boolForKey:kOWS2FAManager_IsEnabledKey
-                            inCollection:kOWS2FAManager_Collection
-                            defaultValue:NO];
+    return self.pinCode != nil;
```
Comment (michaelkirk) [inline SignalServiceKit/src/Util/OWS2FAManager.m]: Thanks for getting rid of this redundant state!

**Pull request merged**